ddtrace/contrib/botocore/services/kinesis.py

```diff
@@ -134,48 +136,79 @@ def patched_kinesis_api_call(original_func, instance, args, kwargs, function_var
-                        )
-                        span.name = schematize_cloud_messaging_operation(
-                            trace_operation,
-                            cloud_provider="aws",
-                            cloud_service="kinesis",
-                            direction=SpanDirection.OUTBOUND,
-                        )
-                except Exception:
-                    log.warning("Unable to inject trace context", exc_info=True)
-
-        try:
-            result = original_func(*args, **kwargs)
-
-            if endpoint_name == "kinesis" and operation == "GetRecords" and config._data_streams_enabled:
-                try:
-                    record_data_streams_path_for_kinesis_stream(pin, params, result, span)
-                except Exception:
-                    log.debug("Failed to report data streams monitoring info for kinesis", exc_info=True)
-
-            set_response_metadata_tags(span, result)
-            return result
-
-        except botocore.exceptions.ClientError as e:
-            # `ClientError.response` contains the result, so we can still grab response metadata
-            set_response_metadata_tags(span, e.response)
-
-            # If we have a status code, and the status code is not an error,
-            #   then ignore the exception being raised
-            status_code = span.get_tag(http.STATUS_CODE)
-            if status_code and not config.botocore.operations[span.resource].is_error_code(int(status_code)):
-                span._ignore_exception(botocore.exceptions.ClientError)
-            raise
+    message_received = False
+    func_run = False
+    func_run_err = None
+    ctx = None
+    start_ns = None
+    result = None
+
+    if config.botocore.propagation_enabled:
+        if endpoint_name == "kinesis" and operation == "GetRecords":
+            try:
+                start_ns = time_ns()
+                func_run = True
+                result = original_func(*args, **kwargs)
+            except Exception as e:
+                func_run_err = e
+            if result is not None and "Records" in result and len(result["Records"]) >= 1:
+                message_received = True
+                ctx = extract_DD_context(result["Records"])
+
+    # we don't want to create a span for empty poll consumer operations
+    if (func_run and message_received) or config.empty_poll_enabled or not func_run:
+        with pin.tracer.start_span(
+            trace_operation,
+            service=schematize_service_name("{}.{}".format(pin.service, endpoint_name)),
+            span_type=SpanTypes.HTTP,
+            child_of=ctx if ctx is not None else pin.tracer.context_provider.active(),
+            activate=True,
+        ) as span:
+            set_patched_api_call_span_tags(span, instance, args, params, endpoint_name, operation)
+
+            if start_ns is not None:
+                span.start_ns = start_ns
```

Comment:
Comment added. I don't think we need to update duration since it is `None` when a span is created and only set once a span is finished. Updating `start_ns` should be the only thing we need AFAIK.

Reply:
My main concern here is that the duration will then include all of the tracing/dsm logic. I guess we already do this in other integrations so it is not a big concern